src/api_schema.rs

```diff
@@ -0,0 +1,138 @@
+                let interface = position.get(schema.schema())?;
+                let remove_children = interface
+                    .fields
+                    .keys()
+                    .map(|field_name| position.field(field_name.clone()))
+                    .collect::<Vec<_>>();
+                for child in remove_children {
+                    child.remove(schema)?;
+                }
+            }
+            position::TypeDefinitionPosition::InputObject(position) => {
+                let input_object = position.get(schema.schema())?;
+                let remove_children = input_object
+                    .fields
+                    .keys()
+                    .map(|field_name| position.field(field_name.clone()))
+                    .collect::<Vec<_>>();
+                for child in remove_children {
+                    child.remove(schema)?;
+                }
+            }
+            position::TypeDefinitionPosition::Enum(position) => {
+                let enum_ = position.get(schema.schema())?;
+                let remove_children = enum_
+                    .values
+                    .keys()
+                    .map(|field_name| position.value(field_name.clone()))
+                    .collect::<Vec<_>>();
+                for child in remove_children {
+                    child.remove(schema)?;
+                }
+            }
+            _ => {}
+        }
+    }
+
+    // TODO is it necessary to remove arguments first?
+    for position in &directives_for_removal {
+        position.remove(schema)?;
+    }
+
+    for position in &types_for_removal {
+        match position {
+            position::TypeDefinitionPosition::Object(position) => {
+                position.remove(schema)?;
+            }
+            position::TypeDefinitionPosition::Interface(position) => {
+                position.remove(schema)?;
+            }
+            position::TypeDefinitionPosition::InputObject(position) => {
+                position.remove(schema)?;
+            }
+            position::TypeDefinitionPosition::Enum(position) => {
+                position.remove(schema)?;
+            }
+            position::TypeDefinitionPosition::Scalar(position) => {
+                position.remove(schema)?;
+            }
+            position::TypeDefinitionPosition::Union(position) => {
+                position.remove(schema)?;
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn to_api_schema(schema: FederationSchema) -> Result<Valid<Schema>, FederationError> {
+    let mut api_schema = schema;
+
+    // As we compute the API schema of a supergraph, we want to ignore explicit definitions of `@defer` and `@stream` because
+    // those correspond to the merging of potential definitions from the subgraphs, but whether the supergraph API schema
+    // supports defer or not is unrelated to whether subgraphs support it.
+    if let Some(defer) = api_schema.get_directive_definition(&name!("defer")) {
```

Comment:
This is how it's done in JS, I think it's specifically relevant for these because the API schema generation *doesn't know* if defer and stream should exist and relies on the router to add them if they should, while the other spec directives always exist and so the shadowing is harmless. At least that's my understanding of it